src/averageAges.js

```diff
@@ -55,10 +72,17 @@ function calculateWomenAverageAge(people, withChildren) {
  * @return {number}
  */
 function calculateAverageAgeDiff(people, onlyWithSon) {
-  // 1. find a mother of each person (or only for men)
-  // 2. keep people who have mothers in the array
-  // 3. calculate the difference child.born - mother.born
-  // 4. return the average value
+  const children = people.filter(({ mother, sex }) => {
+    const isMotherOnlyWithSon = mother && (onlyWithSon && sex === 'm');
+    const personMother = people.find(mom => mother === mom.name);
+
+    return (isMotherOnlyWithSon || (!onlyWithSon)) && personMother;
+  });
+
+  const ageDiff = children.map(({ mother: childMother, born }) =>
+    born - people.find(({ name }) => childMother === name).born);
```

Comment:
For better readability, the search for the mother can be transferred to a separate variable and a ready-made value can be substituted into the calculation.